presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java

```diff
@@ -38,12 +39,24 @@
 
     public static SqlTimestamp fromMillis(int precision, long millis)
     {
-        return newInstance(precision, millis * 1000, 0);
+        return newInstanceWithRounding(precision, millis * 1000, 0);
     }
 
     public static SqlTimestamp newInstance(int precision, long epochMicros, int picosOfMicro)
     {
-        return newInstanceWithRounding(precision, epochMicros, picosOfMicro);
+        if (precision <= 6) {
+            checkArgument(picosOfMicro == 0, "Expected picosOfMicro to be 0 for precision %s: %s", precision, picosOfMicro);
+            checkArgument(round(epochMicros, 6 - precision) == epochMicros, "Expected 0s for digits beyond precision %s: epochMicros = %s", precision, epochMicros);
+        }
+        else {
+            checkArgument(round(picosOfMicro, 12 - precision) == picosOfMicro, "Expected 0s for digits beyond precision %s: picosOfMicro = %s", precision, picosOfMicro);
+        }
+
+        if (picosOfMicro < 0 || picosOfMicro > PICOSECONDS_PER_MICROSECOND) {
+            throw new IllegalArgumentException("picosOfMicro is out of range: " + picosOfMicro);
+        }
```

Comment:
This could use `checkArgument` too

Reply:
Same reasoning either applies or doesn't apply to other checks above, right?

Reply:
Yes, you're right. For some reason, I was thinking that they were in a conditional that was only evaluated sometimes, but that's obviously not the case. I'll revert the changes

Reply:
Yes, you're right. For some reason, I was thinking that they were in a conditional that was only evaluated sometimes, but that's obviously not the case. I'll revert the changes